pynbody/snapshot/ramses.py

```diff
@@ -1143,4 +1155,13 @@ def mass(sim):
 
 @RamsesSnap.derived_quantity
 def temp(sim):
-    return ((sim['p'] / sim['rho']) * (1.22 * units.m_p / units.k)).in_units("K")
+    """ Gas temperature derived from pressure and density """
+    # Has to be redefined and rederived here from Ramses native variables
+    # to avoid running into circular dependencies with the traditional derived definition
+    # Now uses the self-consistent molecular weight field from pynbody (issue 598)
+    from ..derived import mu
+    mu_est = array.SimArray(np.ones(len(sim)), units="1")
+    for i in range(5):
```

Comment:
Why 5?

Reply:
This is matching the existing derived quantity:
https://github.com/pynbody/pynbody/blob/b7bb374404521f3b9922df52dd3d4bd6b2e5ab7e/pynbody/derived.py#L268-L277

I am not entirely sure why 5 iterations were deemed enough or not for the base property.

Reply:
I see... could you fix it in both places? I think there should be a convergence check to a defined tolerance (maybe +/- 0.1%?) and also an upper limit on the number of iterations, with a warning issued if the solution does not converge in time.

Reply:
Sure that sounds good, I can include this this in. Is it ok to do it within this PR or would you want it as a separate fix?

Reply:
Thanks! Fine just to do it here I think.

Reply:
I am having issues with implementing a convergence to a given tolerance. Due to the binary nature of the default mu implementation (either 1.3 or 0.59 depending on a temperature cut to 1e4K):
https://github.com/pynbody/pynbody/blob/b7bb374404521f3b9922df52dd3d4bd6b2e5ab7e/pynbody/derived.py#L235-L250
one cannot actually converge for values around between 5e3 K and 2e4K as the molecular weight just oscillates between 1.3 and 0.59 which each iteration. 

The only solution I see is to improve the actual implementation of `mu` to avoid the binary cut and have a function of temperature that can actually converge. I suggest this is re-opened as an independent issue, as this is global to pynbody rather than just Ramses? 

Any insights?

Reply:
Yes I agree this sounds like it deserves its own issue

Reply:
Done in #605, any other comments on this?